python/pyarrow/_flight.pyx

```diff
@@ -434,19 +473,57 @@ cdef void _do_put(void* self, unique_ptr[CFlightMessageReader] reader):
     descriptor.descriptor = reader.get().descriptor()
     py_reader.reader.reset(reader.release())
     (<object> self).do_put(descriptor, py_reader)
+    return 0
 
 
-cdef void _do_get(void* self, CTicket ticket,
-                  unique_ptr[CFlightDataStream]* stream):
+cdef int _do_get(void* self, CTicket ticket,
+                 unique_ptr[CFlightDataStream]* stream) except -1:
     """Callback for implementing Flight servers in Python."""
-    py_ticket = Ticket()
-    py_ticket.ticket = ticket.ticket
+    py_ticket = Ticket(ticket.ticket)
     result = (<object> self).do_get(py_ticket)
     if not isinstance(result, FlightDataStream):
         raise TypeError("FlightServerBase.do_get must return "
                         "a FlightDataStream")
     stream[0] = move((<FlightDataStream> result).stream)
+    return 0
+
+
+cdef int _do_action_result_next(void* self,
+                                unique_ptr[CResult]* result) except -1:
+    """Callback for implementing Flight servers in Python."""
+    try:
+        action_result = next(<object> self)
+        if not isinstance(action_result, Result):
+            raise TypeError("Result of FlightServerBase.do_action must "
+                            "return a Result during iteration")
```

Comment:
I would say "FlightServerBase.do_action must yield Result objects" or "FlightServerBase.do_action must return an iterator of Result objects".